cachi2/core/models/sbom.py

```diff
@@ -104,3 +115,585 @@ class Sbom(pydantic.BaseModel):
+    """
+
+    annotator: str
+    annotationDate: str
+    annotationType: Literal["OTHER", "REVIEW"]
+    comment: str
+
+    def __hash__(self) -> int:
+        return hash((self.annotator, self.annotationDate, self.annotationType, self.comment))
+
+
+class SPDXChecksum(pydantic.BaseModel):
+    """A basic representation of a checksum entry."""
+
+    algorithm: str
+    checksumValue: str
+
+    def __hash__(self) -> int:
+        return hash(self.algorithm + self.checksumValue)
+
+
+class SPDXFile(pydantic.BaseModel):
+    """SPDX File.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/
+
+    An actual SPDX document generated from a directory or a container image
+    would usually contain "files" section. This section would contain file
+    entries of this shape.
+    """
+
+    SPDXID: Optional[str] = None
+    fileName: str
+    checksums: list[SPDXChecksum] = []  # TODO: flesh out proper cehcksums type
+    licenseConcluded: str = "NOASSERTION"
+    copyrightText: str = ""
+    comment: str = ""
+
+    def __hash__(self) -> int:
+        return hash(
+            hash(self.SPDXID)
+            + hash(self.fileName)
+            + hash(self.licenseConcluded + self.copyrightText + self.comment)
+            + sum(hash(c) for c in self.checksums)
+        )
+
+
+class SPDXPackage(pydantic.BaseModel):
+    """SPDX Package.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/
+    """
+
+    SPDXID: Optional[str] = None
+    name: str
+    versionInfo: Optional[str] = None
+    externalRefs: list[SPDXPackageExternalRefType] = []
+    annotations: list[SPDXPackageAnnotation] = []
+    downloadLocation: str = "NOASSERTION"
+
+    def __lt__(self, other: "SPDXPackage") -> bool:
+        return (self.SPDXID or "") < (other.SPDXID or "")
+
+    def __hash__(self) -> int:
+        return hash(
+            hash(self.SPDXID)
+            + hash(self.name)
+            + hash(self.versionInfo)
+            + hash(self.downloadLocation)
+            + sum(hash(e) for e in self.externalRefs)
+            + sum(hash(a) for a in self.annotations)
+        )
+
+    @staticmethod
+    def _calculate_package_hash_from_dict(package_dict: Dict[str, Any]) -> str:
+        return hashlib.sha256(json.dumps(package_dict, sort_keys=True).encode()).hexdigest()
+
+    @pydantic.field_validator("externalRefs")
+    def _purls_validation(
+        cls, refs: list[SPDXPackageExternalRefType]
+    ) -> list[SPDXPackageExternalRefType]:
+        """Validate that SPDXPackage includes only one purl with the same type, name, version."""
+        purls = [ref.referenceLocator for ref in refs if ref.referenceType == "purl"]
+        parsed_purls = [PackageURL.from_string(purl) for purl in purls if purl]
+        unique_purls_parts = set([(p.type, p.name, p.version) for p in parsed_purls])
```

Comment:
I'm not sure if treating purls as equal (or "similar enough") just by `type, name, version` (or `type, namespace, name, version`) is a good idea.

Cachi2 takes care to report "non-standard" dependencies as such by adding qualifiers and/or subpaths.

```
pkg:npm/my-package@1.0.0
pkg:npm/my-package@1.0.0?vcs_url=git+https://github.com/some-org/my-package@deadbeef#my-package
```
^ the first is a dependency from registry.npmjs.org, the second is from a subdirectory of a github repo. There's no guarantee that the two are related by anything other than coincidence

What are the intended semantics of having one package with multiple purls? Is it that you can pick any one of them arbitrarily and it will represent the same package? If so, that's currently not true

Reply:
The code at L394 appears to be validating a single package's list of purls. Having two unrelated purls that happen to share type, name and  version within one package looks like a bug (and is reported as a run time issue on L396). Note, that this code does not merge individual packages, that happens further down around L503. The docstring says:

>If package with same name and version is found multiple times in the list,
>merge external references of all the packages into one package.

This could result in merging of two unrelated packages together (like in your example).

> There's no guarantee that the two are related by anything other
> What are the intended semantics  than coincidence...? 

My understanding of the semantics is that a list of purls represents all versions of a certain package (distinguishable by `type, name, version` tuple) that were used to generate certain artifact and either are present within it, or were present at some point. In the case when there are multiple unrelated (i.e. different), but identically named and versioned packages it would be currently impossible to distinguish between them and picking one at random would be a mistake. Is this a normal and expected situation?